pkg/apis/integreatly/v1alpha1/rhmiconfig_types.go

```diff
@@ -101,17 +90,13 @@ type Upgrade struct {
 	// contacts: list of contacts which are comma separated
 	// "user1@example.com,user2@example.com"
 	Contacts string `json:"contacts,omitempty"`
-	// always-immediately: boolean value, if set to true an upgrade will be applied as soon as it is available,
-	// whether service affecting or not.
-	// This takes precedence over all other options
-	AlwaysImmediately bool `json:"alwaysImmediately"`
-	// during-next-maintenance: boolean value, if set to true an upgrade will be applied within the next maintenance window.
-	// Takes precedence over apply-on
-	DuringNextMaintenance bool `json:"duringNextMaintenance"`
-	// apply-on: string date value. If 'always-immediately' or 'during-next-maintenance' is not set the customer is
-	// required to pick a time for the upgrade. Time value will be validated by a webhook and reset to blank after
-	// upgrade has completed. Format: "dd MMM YYYY hh:mm" > "12 Jan 1980 23:00". UTC time
-	ApplyOn string `json:"applyOn,omitempty"`
+
+	// If this value is true, upgrades will be approved in the next maintenance window
+	// n days after the upgrade is made available. Being n the value of `notBeforeDays`.
+	WaitForMaintenance string `json:"waitForMaintenance,omitempty"`
+
+	// Minimum of days since an upgrade is made available until it's approved
+	NotBeforeDays string `json:"notBeforeDays,omitempty"`
```

Comment:
Is it because of default values?
Maybe we will be able to use it soon - https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#defaulting

Reply:
Yes, it's for support of default values. Specifying the type as integer as boolean would mean that when the values aren't specified we would get 0 and false respectively, which is a totally acceptable value the user might want to set, so there's no other way to know when we get the zero value if it was specified by the user or non present.
Also for the upgrade from the previous CR in for existing installations, if the user doesn't manually set the new fields after updating, the operator would only see that the value is 0 and false, so a subsequent upgrade would be applied immediately

Reply:
We are controlling how RHMIConfig gets created, so we could set defaults then.
If we can come up with some solution for the upgrade of old CR, would it be worth it to switch to boolean/integer types?

Reply:
Yes. It's always better to deal directly with the typed values, more errors that the compiler handles instead of us!
I could experiment with pointers so the zero value would be `nil` and we still have the types. I'm not sure if k8s supports this but it's worth a try

Reply:
I think the old fields might be preserved when upgrading to new CRD.
So maybe we can check if they are present, set the default and remove them?

Reply:
The old fields are presented in the yaml representation of the CR, but when unmarshalling to the Go type they will simply be ignored, so when the spec is:
```yaml
upgrade:
  alwaysImmediately: false
  duringMaintenance: false
```

The Go value will be:
```go
Upgrade{
  WaitForMaintenance: false
  NotBeforeDays: 0
}
```

Therefore there's no way (without directly handling the json) for the operator to know when the other fields are still present.

Reply:
I pushed a commit to change the types to `*int` and `*bool`, this is handled correctly so when the value is not present it's set to `nil` so it's just a matter of checking this to know when to use the default. Then once the CR is updated the old fields are automatically removed so it's all pretty seamless

Reply:
Nice approach, worked well.